### ISSUE TITLE
Use shaders instead of Gdscript for HSV adjustments

### ISSUE THREAD

**Issue by anasofgo** [Orama-Interactive/Pixelorama#343]
I noticed that the performance drops significantly when you have a large picture you are working on. I tried to make a 1000x1000 art work (which is not that big) and if I try filling it with color the program freezes for a few seconds.. I then try adjusting the HSV from the "Image" menu and the program crashes.

I tried to dig up the issue and I found the following function:

```
func adjust_hsv(img: Image, delta_h : float, delta_s : float, delta_v : float, pixels : Array) -> void:
	img.lock()
	for i in pixels:
		var c : Color = img.get_pixelv(i)
		# Hue
		var hue = range_lerp(c.h,0,1,-180,180)
		hue = hue + delta_h

		while(hue >= 180):
			hue -= 360
		while(hue < -180):
			hue += 360

		# Saturation
		var sat = c.s
		if delta_s > 0:
			sat = range_lerp(delta_s,0,100,c.s,1)
		elif delta_s < 0:
			sat = range_lerp(delta_s,-100,0,0,c.s)

		# Value
		var val = c.v
		if delta_v > 0:
			val = range_lerp(delta_v,0,100,c.v,1)
		elif delta_v < 0:
			val = range_lerp(delta_v,-100,0,0,c.v)

		c.h = range_lerp(hue,-180,180,0,1)
		c.s = sat
		c.v = val
		img.set_pixelv(i,c)

	img.unlock()

```
This for loop is apparently very taxing in terms of computational power, and it is also used for the "preview" window.

If there is no more efficient workaround solution for this, I think it would be wise to make it be called out only when the user confirms his adjustments. it shouldn't be used to update the little "preview" window.

For the preview, I suggest that a simple hsv adjustment shader is used. Shaders are faster and more efficient for "realtime" previews.

Another additional thing that I found interesting is this https://github.com/Coldragon/godot-shader-to-image . Essentially a shader can be turned into an image, so it is possible to add extra functionality and effects more easily using shaders. I am still experimenting with this approach.

The following are my Laptop specs:
Nvidia Gtx 1060
intel Core i7
16 GB ram
Windows 10

**Comment by luiq54** [comment]
HSV adjustment dialog has been moved to shader based one in a [PR](https://github.com/Orama-Interactive/Pixelorama/pull/475). Can you try to get a build from [here](https://github.com/Orama-Interactive/Pixelorama/actions/runs/767494430) and compare results ?

**Comment by anasofgo** [comment]
@luiq54  I can confirm that it is working properly, I didnt experience any crashes with the new build..Thanks for the PR! 
One suggestion: Although the crashing problem has been solved, adjusting the sliders is still a bit laggy on higher image sizes, would it be a good idea to scale down the preview texture by a certain percentage for higher resolution images to make the processing faster? Or maybe have two options for preview: Fast preview and Accurate preview.

**Comment by luiq54** [comment]
That is a really cool Idea!! I will see if it is possible. Thanks! 😄

**Comment by luiq54** [comment]
I have created a PR for same. You can access the builds from [here](https://github.com/Orama-Interactive/Pixelorama/actions/runs/1042212667).

I am not sure whether this will be beneficial as the previews are updated frequently and the process to shrink preview_images is also very intensive. This can be simplified but will require major rewrites.

**Comment by Variable-ind** [comment]
I just created a variant of your **Fast preview and Accurate preview** idea [Here](https://github.com/Variable-ind/Pixelorama/tree/Boast_HSV_and_Rotation). please check it out and tell me what you think about it :smiley:

**Comment by anasofgo** [comment]
Thanks thats great, I tried it and it definitely improved the user experience!
In my opinion, it is great that we can give the user the option to control the time to update the preview window.

I have a small suggestion that might help with the usability, rather than having to write a number each time the user switches between images of different sizes, maybe a small checkbox for "Live update for preview" can be placed. If the user ticks the box then the preview will update instantly as the user moves the slider, of course the user can also adjust the wait time if they wish.. If the "Live update for preview" is not ticked, then the preview will only update once the user releases the mouse button from the sliders.
This suggestion has the following advantage in my opinion:
As stated before, the user no longer needs to worry about changing the wait time value when they switch between 2 files with different resolutions (most of the time, they wont need to have live updates if the resolution is big because that would cause a lag), a simple checkbox will take care of making the preview Realtime or not.

**Comment by Variable-ind** [comment]
I made the changes [Here!](https://github.com/Variable-ind/Pixelorama/tree/Boast_HSV_and_Rotation). But at this point i can think of two different implementations...
1) The delay preview method as mentioned above...
2) A manual **Update preview** button that comes visible when **Live preview** is unchecked (Implementation given [Here](https://github.com/Variable-ind/Pixelorama/tree/Boost_HSV_implement_2))

There **is** an obvious downside to the second option though that the user would have to press **Update preview** every time he wants to update the preview.

Try out both and tell me what what you suggest :smiley:

**Comment by anasofgo** [comment]
Thanks for the changes 😃 , I tried both and I think the first implementation is the better one (you actually did it better than I had described). As you said, expecting the user to press "Update preview" each time is a big downside, I think most users would like to see the preview updated as fast as possible without having a lag (especially if the users come from other software where updates are instantaneous). The first implementation addresses the issue quite well and improves the usability to an acceptable level.
